lib/ansible/plugins/httpapi/fortimanager.py

```diff
@@ -0,0 +1,348 @@
+
+        if "FortiManager object connected to FortiManager" in self.__str__():
+            # If Login worked, then inspect the FortiManager for Workspace Mode, and it's system information.
+            self.inspect_fmgr()
+            return
+        else:
+            raise FMGBaseException(msg="Unknown error while logging in...connection was lost during login operation...."
+                                       " Exiting")
+
+    def inspect_fmgr(self):
+        # CHECK FOR WORKSPACE MODE TO SEE IF WE HAVE TO ENABLE ADOM LOCKS
+        self._lock_ctx.check_mode()
+        # CHECK FOR SYSTEM STATUS -- SHOULD RETURN 0
+        status = self.get_system_status()
+        if status[0] == -11:
+            # THE CONNECTION GOT LOST SOMEHOW, REMOVE THE SID AND REPORT BAD LOGIN
+            self.logout()
+            raise FMGBaseException(msg="Error -11 -- the Session ID was likely malformed somehow. Contact authors."
+                                       " Exiting")
+        elif status[0] == 0:
+            try:
+                self._connected_fmgr = status[1]
+                self._host = self._connected_fmgr["Hostname"]
+            except BaseException:
+                pass
+        return
+
+    def logout(self):
+        """
+        This function will logout of the FortiManager.
+        """
+        if self.sid is not None:
+            if self._lock_ctx.uses_workspace:
+                self._lock_ctx.run_unlock()
+            ret_code, response = self.execute("sys/logout")
+            self.sid = None
+            self.connection = None
+            return ret_code, response
+
+    def send_request(self, method, params, login=False):
+        """
+        Responsible for actual sending of data to the connection httpapi base plugin. Does some formatting as well.
+        :param params: A formatted dictionary that was returned by self.common_datagram_params()
+        before being called here.
+        :param method: The preferred API Request method (GET, ADD, POST, etc....)
+        :type method: basestring
+        :param login: Tells the function that we're attempting a login, and that the self.sid should be empty,
+        but to continue anyway. Do not use this unless being called from self.login().
+
+        :return: Dictionary of status, if it logged in or not.
+        """
+        if self.sid is None and not login:
+            raise FMGValidSessionException(method, params)
+        self._update_request_id()
+        json_request = {
+            "method": method,
+            "params": params,
+            "session": self.sid,
+            "id": self.req_id,
+            "verbose": 1
+        }
+        data = json.dumps(json_request, ensure_ascii=False).replace('\\\\', '\\')
+        url = str(self._url)
+        try:
+            # Sending URL and Data in Unicode, per Ansible Specifications for Connection Plugins
+            response, response_data = self.connection.send(path=to_text(url), data=to_text(data),
```

Comment:
`self._url` is hardcoded (to `"/jsonrpc"`) in `login()`, sent through `to_text()` here, and also stringified on the immediately preceding line of executing code. on the next line of executing code. Is there any particular reason for all this, or would it be just as clear to do this instead?
```suggestion
            response, response_data = self.connection.send(path='/jsonrpc', data=to_text(data),
```

Reply:
I've hardcoded this under _init_ and simply referenced it directly in cases where it used.